### PR TITLE
Add TabEBM implementation for data augmentation using SGLD

### PR DESCRIPTION
**Pull request by SilenceX12138**
This pull request introduces the new TabEBM extension to the project, which provides a high-performance method for tabular data augmentation using class-specific energy-based models. The README and project documentation are updated to include TabEBM, and a comprehensive new README for TabEBM is added, detailing its features, installation, and usage.

**Documentation and Extension Integration:**

* Added TabEBM as a new extension for data augmentation in the main project documentation, including a brief description and integration into the module list (`README.md`).
* Updated the project architecture diagram and related sections to include TabEBM as an unsupervised task, along with a direct link to an example notebook (`README.md`). [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R107) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L168-R170) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R181)

**TabEBM Documentation:**

* Added a comprehensive new `README.md` for TabEBM, describing its purpose, features (such as optimized implementation, GPU acceleration, and memory-efficient sampling), installation instructions, quick start guide, and links to interactive tutorials.

### CODE REVIEW

**Comment by SilenceX12138** [comment]
> Hi @SilenceX12138, thank you so much this extensive PR, we would love to have this in the package. It appears that right now some of your new test cases are failing. Once these are passing, ping me again and I will review the entire PR :)

Hi @klemens-floege. I have updated the code format, and it passes all the `ruff` checks on my local machine. Could you please check out the latest version when it works? Thank you!

**Comment by SilenceX12138** [comment]
Hi @klemens-floege. I have resolved the compatibility issues related to NumPy’s behaviour across versions and platforms. When you have some time, could you please test the latest changes and let me know if everything works on your end? Thanks!

File: examples/tabebm/tabebm_augment_real_world_data.ipynb
Comment (klemens-floege) [inline examples/tabebm/tabebm_augment_real_world_data.ipynb:1]: I like this Notebook and got it to run on my end. I have 2 comments: 

- While Notebooks are great for showcasing your methods, they are not ideal for being checked into git + maintained. Since the code does not have visualizations, I think it would be best if we move this code into a single example.py file. 
-You are using tabcamel to load the example data, while this library is quite neat, it seems a bit unnecessary to add a dependency for this. Would it be possible to just use fetch_openml, as in the large_datasets_example.py file. The adult dataset should also be available under ID 45068 I believe.

File: src/tabpfn_extensions/tabebm/tabebm.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+        Args:
+            X: Input features tensor
+            y: Input labels tensor
+
+        Returns:
+            Dictionary containing prepared batch data
+        """
+        # Use efficient train/test split with no validation data
+        splitter = partial(
+            TabEBM.train_test_split_allow_full_train,
```
Comment (klemens-floege) [inline src/tabpfn_extensions/tabebm/tabebm.py:435]: I you want to use all data for train can you not just skip the call entirely? In there case you can use sklearns train_test_split

File: src/tabpfn_extensions/tabebm/tabebm.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        X: np.ndarray | torch.Tensor | pd.DataFrame,
+        y: np.ndarray | torch.Tensor | pd.Series,
+        num_samples: int,
+        starting_point_noise_std: float = 0.01,
+        sgld_step_size: float = 0.1,
+        sgld_noise_std: float = 0.01,
+        sgld_steps: int = 200,
+        distance_negative_class: float = 5,
+        seed: int = 42,
+        debug: bool = False,
```
Comment (klemens-floege) [inline src/tabpfn_extensions/tabebm/tabebm.py]: It's better not to have these debug flags lying around. In case you would like to support a setting where we print things during sampling etc. rename to verbose

File: src/tabpfn_extensions/tabebm/tabebm.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+        X: np.ndarray,
+        y: np.ndarray,
+        num_samples: int,
+        starting_point_noise_std: float = 0.01,
+        sgld_step_size: float = 0.1,
+        sgld_noise_std: float = 0.01,
+        sgld_steps: int = 200,
+        distance_negative_class: float = 5,
+        seed: int = 42,
+        debug: bool = False,
```
Comment (klemens-floege) [inline src/tabpfn_extensions/tabebm/tabebm.py]: Here I would remove debug since either just skips this sampling step, in case we ever want to skip this then lets have another parameter that explicitly states this

File: src/tabpfn_extensions/tabebm/tabebm.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+
+        # Pre-allocate noise tensor for all SGLD steps (memory optimization)
+        sample_shape = (num_samples, X.shape[1])
+        noise_shape = (sgld_steps, *sample_shape)
+
+        # Set seed for reproducible noise generation
+        seed_everything(seed)
+
+        for target_class in unique_classes:
+            if debug:
```
Comment (klemens-floege) [inline src/tabpfn_extensions/tabebm/tabebm.py]: This is already done above, but as per comments above should probs remove these

File: src/tabpfn_extensions/tabebm/tabebm.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+        for t in range(sgld_steps):
+            # Clear previous gradients
+            if X_sgld_list[0].grad is not None:
+                X_sgld_list[0].grad.zero_()
+
+            # Forward pass to compute energy
+            logits = self.model.forward(X_sgld_list, return_logits=True)
+
+            # Compute energy and total loss
+            energy = TabEBM.compute_energy(logits.reshape(logits.shape[-1], -1))
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/tabebm/tabebm.py:508]: I tested this on my end and got exactly the same results as before @SilenceX12138 what do you think?

```
Vanilla model's balanced accuracy: 71.33
Augmented model's balanced accuracy: 75.99
```

File: examples/tabebm/tabebm_augment_real_world_data.ipynb
Comment (klemens-floege) [reply examples/tabebm/tabebm_augment_real_world_data.ipynb:1]: On a slightly different note I generated 100 samples and got even better results which is great. 

Augmented model's balanced accuracy: 75.99 -> Augmented model's balanced accuracy: 77.60

File: src/tabpfn_extensions/tabebm/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+[![Test In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/andreimargeloiu/TabEBM/blob/main/tutorials/tutorial3_augment_real_world_data.ipynb)
+[![Python 3.10+](https://img.shields.io/badge/python-3.10+-blue.svg)](https://www.python.org/downloads/release/python-3100/)
+[![Downloads](https://static.pepy.tech/badge/tabebm)](https://pypi.org/project/tabebm/)
+
+</div>
+
+Official code for the paper ["TabEBM: A Tabular Data Augmentation Method with Distinct Class-Specific Energy-Based Models"](https://arxiv.org/abs/2409.16118), published in the Thirty-Eighth Annual Conference on Neural Information Processing Systems (NeurIPS 2024).
+
+Authored by [Andrei Margeloiu\*](https://www.linkedin.com/in/andreimargeloiu/), [Xiangjian Jiang\*](https://silencex12138.github.io/), [Nikola Simidjievski](https://simidjievskin.github.io/), [Mateja Jamnik](https://www.cl.cam.ac.uk/~mj201/), University of Cambridge, UK
+
```
Comment (klemens-floege) [inline src/tabpfn_extensions/tabebm/README.md:21]: Would be great to list you as primary maintainer as well

```suggestion

**Primary Maintainer:** Xiangjian Jiang

```

File: src/tabpfn_extensions/tabebm/tabebm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SilenceX12138) [reply src/tabpfn_extensions/tabebm/tabebm.py:508]: Hi @klemens-floege. Thank you for the review! However, there are a few key points where the current suggestion appears to be based on an incorrect assumption:

**1. Incorrect Assumption About Tensor Shape**
The comment assumes that the shape of logits is `(1, num_samples, num_classes)`, but in fact, the correct shape is `(1, num_classes, num_samples)`. This distinction is critical because reshaping from `(1, C, N)` to `(N, C)` using `logits.reshape(logits.shape[-1], -1)` yields the intended (num_samples, num_classes) shape expected by compute_energy.

**2. Misdiagnosis Based on Static Analysis**
This feedback appears to rely solely on static shape analysis, likely inferred textually or by an LLM. While helpful in many cases, such methods are prone to subtle misinterpretations, especially when tensor semantics are involved. I would recommend verifying tensor shapes at runtime using breakpoints or print statements to avoid incorrect assumptions. :)))

**3. Behaviour Appears Correct Due to Binary Surrogate Setup**
The results may appear consistent because TabEBM currently applies a binary surrogate task, which masks shape-related inconsistencies (i.e., the logit dimension is always 1, leading to equivalent `reshape`/`permute`/`squeeze`). However, the current implementation is deliberately structured to support more general settings, and any changes here could break behaviour in multiclass scenarios or under structural assumptions used elsewhere in the model.

For these reasons, I suggest retaining the existing reshaping logic as it correctly aligns with the input expectations of `compute_energy` and supports the broader design of the model.

File: src/tabpfn_extensions/tabebm/tabebm.py (hunk 1 above)
Comment (SilenceX12138) [reply src/tabpfn_extensions/tabebm/tabebm.py:435]: 1. You might consider trying `train_test_split`, though it appears not to support a full training setup (i.e., using the entire dataset with `test_size=0`).
2. Additionally, the `split_fn` argument is mandatory for TabPFN’s `get_preprocessed_datasets`. Skipping it would require more fundamental changes to the TabPFN model itself. If you're interested in modifying it, I’d be glad to see TabPFN become more flexible in this regard.

Given these constraints, I’d recommend keeping the current code unchanged.

File: src/tabpfn_extensions/tabebm/tabebm.py (hunk 2 above)
Comment (SilenceX12138) [reply src/tabpfn_extensions/tabebm/tabebm.py]: Renamed

File: examples/tabebm/tabebm_augment_real_world_data.ipynb
Comment (SilenceX12138) [reply examples/tabebm/tabebm_augment_real_world_data.ipynb:1]: Added a script as requested

**Comment by SilenceX12138** [comment]
Hi @klemens-floege, thanks for flagging this. I’ve double-checked on my local machine, and the results seem consistent there. The difference you’re seeing is likely due to hardware variances rather than an implementation issue. On my end, I can robustly reproduce the results shown below.

It’s also worth noting that the method is not expected to consistently outperform the original data in every run. The improvements reported in the paper are based on averages across multiple splits, rather than on a single run.

In short, I’d suggest keeping the code as it is.


```shell
Loading adult dataset from OpenML...
Dataset shape: (48842, 14)
Target classes: ['<=50K' '>50K']
Feature names: ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week', 'workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
Subsampling dataset to 200 samples...
Subsampled dataset shape: (200, 14)
Class distribution: class
<=50K    152
>50K      48
Name: count, dtype: int64
Training set size: 160
Test set size: 40
Categorical features: ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
Numerical features: ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']
Processed training data shape: (160, 68)
Processed test data shape: (40, 68)
Unique target labels: [0 1]

==================================================
GENERATING SYNTHETIC DATA WITH TabEBM
==================================================
Original training set size: 160
Synthetic data size: 100
Augmented training set size: 260

==================================================
TRAINING DOWNSTREAM PREDICTORS
==================================================

Training vanilla model on original data...
Training augmented model on real + synthetic data...

==================================================
EVALUATION RESULTS
==================================================
Vanilla model's balanced accuracy: 66.67%
Augmented model's balanced accuracy: 68.33%
Improvement: 1.67 percentage points
✓ Data augmentation with TabEBM improved model performance!
```
